lib/auth/in_memory/AuthLoader.js

```diff
@@ -0,0 +1,231 @@
+            .required();
+
+        let sasToken = joi.string();
+        if (checkSas) {
+            sasToken = sasToken.required();
+        }
+        let accountsJoi = joi.array()
+                .items({
+                    name: joi.string().required(),
+                    email: joi.string().email().required(),
+                    arn: joi.string().required(),
+                    canonicalID: joi.string().required(),
+                    shortid: joi.string().regex(/^[0-9]{12}$/).required(),
+                    sasToken,
+                    keys: this._joiKeysValidator,
+                    // backward-compat
+                    users: joi.array(),
+                })
+                .required()
+                .unique('arn')
+                .unique('email')
+                .unique('canonicalID');
+        if (checkSas) {
+            accountsJoi = accountsJoi.unique('sasToken');
+        }
+        this._joiValidator = joi.object({ accounts: accountsJoi });
+    }
+
+    /**
+     * add one or more accounts to the authentication info
+     *
+     * @param {object} authData - authentication data
+     * @param {object[]} authData.accounts - array of account data
+     * @param {string} authData.accounts[].name - account name
+     * @param {string} authData.accounts[].email: email address
+     * @param {string} authData.accounts[].arn: account ARN,
+     *   e.g. 'arn:aws:iam::123456789012:root'
+     * @param {string} authData.accounts[].canonicalID account
+     *   canonical ID
+     * @param {string} authData.accounts[].shortid account ID number,
+     *   e.g. '123456789012'
+     * @param {string} authData.accounts[].sasToken SAS token for
+     *   Azure
+     * @param {object[]} authData.accounts[].keys array of
+     *   access/secret keys
+     * @param {object[]} authData.accounts[].keys[].access access key
+     * @param {object[]} authData.accounts[].keys[].secret secret key
+     * @param {string} [filePath] - optional file path info for
+     *   logging purpose
+     * @return {undefined}
+     */
+    addAccounts(authData, filePath) {
+        const isValid = this._validateData(authData, filePath);
+        if (isValid) {
+            this._authData.accounts =
+                this._authData.accounts.concat(authData.accounts);
+            // defer validity checking when getting data to avoid
+            // logging multiple times the errors (we need to validate
+            // all accounts at once to detect duplicate values)
+            if (this._isValid) {
+                this._isValid = null;
+            }
+        } else {
+            this._isValid = false;
+        }
+    }
+
+    /**
+     * add account information from a file
+     *
+     * @param {string} filePath - file path containing JSON
+     *   authentication info (see {@link addAccounts()} for format)
+     * @return {undefined}
+     */
+    addFile(filePath) {
+        const authData = JSON.parse(fs.readFileSync(filePath));
```

Comment:
do you need a try/catch wrapper?

Reply:
For now, it should be ok because we already throw when the auth config is wrong (which is fine since we don't want to start the process in such case, and the auth config is all loaded at startup). But for being more robust we may want to catch those errors at some point, at this point I don't know where we want to do this, it will depend on actual future use of the class as a dynamic loader, so for now I prefer to leave as is.